File: recipes/list.py
from __future__ import annotations
from re import sub

from utils.loader import GobiFile
from utils.recipes import GobiError, Action, Recipe


class ListAction(Action):
    def __init__(self) -> None:
        super().__init__()
        self.name = "list"
        self.subname = "list"

    def help(self) -> str:
        return """
Get a list of all available actions for a project
Usage: gobi <project list...>? list [porcelain|full]

By default will try and display the subname of available actions unless a conflict exists, then will use the full name.
Porcelain will not print out the message at the start of the list command.
Full will print out the full name of all actions.
"""

    @staticmethod
    def print_list(actions: list[Action], full: bool = False) -> None:
        if full:
            for action in sorted(actions, key=lambda a: a.subname):
                print(action.name)
        else:
            minimal_name: dict[str, list[Action]] = {}
            for action in actions:
                if action.subname not in minimal_name:
                    minimal_name[action.subname] = [action]
                else:
                    minimal_name[action.subname].append(action)
            sorted_subnames = sorted(minimal_name.keys())
            for subname in sorted_subnames:
                actions = minimal_name[subname]
                priority_actions = [a for a in actions if a.priority]
                if len(priority_actions) == 1:
                    print(subname)
                elif len(priority_actions) > 1:
                    for a in sorted(priority_actions, key=lambda a: a.name):
                        print(a.name)
                elif len(actions) == 1:
                    print(subname)
                else:
                    for a in sorted(actions, key=lambda a: a.name):
                        print(a.name)

    def run(
        self,
        gobi_file: GobiFile,
        recipes: dict[str, Recipe],
        actions: list[Action],
        args: list[str],
    ) -> GobiError | None:
        match args:
            case []:
                print("List of available actions:")
                ListAction.print_list(actions)
            case ["porcelain"]:
                ListAction.print_list(actions)
            case ["full"]:
                ListAction.print_list(actions, full=True)
            case [arg]:
                return GobiError(self, 1, f"Unknown argument: {arg}")
            case _:
                return GobiError(self, 1, "Too many arguments passed to list action")


class ListRecipe(Recipe):
    def __init__(self):
        self.name = "list"

    def help(self) -> str:
        return "Generates the list action for gobi."

    def create_actions(self, gobi_file: GobiFile) -> GobiError | tuple[list[Action], list[str]]:
        return [ListAction()], []


def create() -> Recipe:
    return ListRecipe()

File: lib/recipes/src/list.rs
use gobi_lib::{file::GobiFile, recipes::*, GobiError, GobiResult};

pub struct ListAction {}

impl ListAction {
    pub fn print_list(actions: &Vec<Action>, full: bool) {
        let mut actions: Vec<Action> = actions.clone();
        if full {
            actions.sort_by(|a, b| a.get_name().cmp(b.get_name()));
            actions.iter().for_each(|action| {
                println!("{}", action.get_name());
            });
        } else {
            get_minimal_names(&actions).iter().for_each(|(_, name)| {
                println!("{}", name);
            });
        }
    }
}

impl IAction for ListAction {
    fn get_name(&self) -> &str {
        "list"
    }

    fn get_subname(&self) -> &str {
        "list"
    }

    fn get_help(&self) -> &str {
        "Get a list of all available actions for a project
Usage: gobi <project list...>? list [porcelain|full]

By default will try and display the subname of available actions unless a conflict exists, then will use the full name.
Porcelain will not print out the message at the start of the list command.
Full will print out the full name of all actions."
    }

    fn run(&self, actions: &Vec<Action>, args: Vec<String>) -> GobiResult<()> {
        if args.len() == 0 {
            println!("Available actions:");
            ListAction::print_list(actions, false);
        } else if args.len() == 1 {
            match args[0].as_str() {
                "porcelain" => ListAction::print_list(actions, false),
                "full" => ListAction::print_list(actions, true),
                _ => {
                    return Err(GobiError {
                        code: 1,
                        msg: format!("Unknown argument: {}", args[0]),
                    });
                }
            }
        } else {
            return Err(GobiError {
                code: 1,
                msg: "Too many arguments passed to list action".to_string(),
            });
        }
        return Ok(());
    }
}

struct ListRecipe {}

impl ListRecipe {
    fn new() -> ListRecipe {
        ListRecipe {}
    }
}

impl IRecipe for ListRecipe {
    fn get_name(&self) -> &str {
        "list"
    }

    fn get_help(&self) -> &str {
        "Generates a list of all available actions"
    }

    fn create_actions(
        &self,
        _recipe_manager: &SharedRecipeManager,
        _gobi_file: &GobiFile,
    ) -> GobiResult<Vec<Action>> {
        return Ok(vec![ActionWrapper::new(ListAction {})]);
    }
}

pub fn register_recipes(recipe_manager: &mut RecipeManager) {
    recipe_manager.register_recipe(RecipeWrapper::new(ListRecipe::new()));
}
